bucket-registry.go

```diff
@@ -29,54 +29,117 @@ import (
 // `realpath` or `fcntl(F_GETPATH)`) and use the canonical path as the key.
 // Unfortunately Go doesn't seem to have an API for that.
 
-var bucketRegistry = map[string][]*Bucket{} // Maps URL to slice of Buckets at that URL
-var bucketRegistryMutex sync.Mutex          // Thread-safe access to bucketRegistry
+type bucketRegistry struct {
+	byURL       map[string][]*Bucket
+	inMemoryRef map[string]*Bucket
+	lock        sync.Mutex
+}
 
-// Adds a newly opened Bucket to the registry.
-func registerBucket(bucket *Bucket) {
+var cluster *bucketRegistry // global cluster registry
+func init() {
+	cluster = &bucketRegistry{
+		byURL:       make(map[string][]*Bucket),
+		inMemoryRef: make(map[string]*Bucket),
+	}
+}
+
+// registryBucket adds a newly opened Bucket to the registry.
+func (r *bucketRegistry) registerBucket(bucket *Bucket) {
 	url := bucket.url
-	if isInMemoryURL(url) {
-		return
+	name := bucket.GetName()
+	debug("registerBucket %v %s at %s", bucket, name, url)
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	_, ok := r.inMemoryRef[name]
+	if !ok {
+		b := bucket.copy()
+		r.inMemoryRef[name] = b
 	}
-	debug("registerBucket %v at %s", bucket, url)
-	bucketRegistryMutex.Lock()
-	bucketRegistry[url] = append(bucketRegistry[url], bucket)
-	bucketRegistryMutex.Unlock()
+	r.byURL[url] = append(r.byURL[url], bucket)
+}
+
+func (r *bucketRegistry) getInMemoryBucket(name string) *Bucket {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.inMemoryRef[name]
 }
 
 // Removes a Bucket from the registry. Must be called before closing.
-func unregisterBucket(bucket *Bucket) {
+func (r *bucketRegistry) unregisterBucket(bucket *Bucket) {
 	url := bucket.url
-	if isInMemoryURL(url) {
-		return
-	}
-	debug("UNregisterBucket %v at %s", bucket, url)
-	bucketRegistryMutex.Lock()
-	defer bucketRegistryMutex.Unlock()
+	name := bucket.name
+	debug("UNregisterBucket %v at %s", bucket, name, url)
+	r.lock.Lock()
+	defer r.lock.Unlock()
 
-	buckets := bucketRegistry[url]
+	buckets := r.byURL[url]
 	i := slices.Index(buckets, bucket)
 	if i < 0 {
 		warn("unregisterBucket couldn't find %v", bucket)
 		return
 	}
 	if len(buckets) == 1 {
-		delete(bucketRegistry, url)
-	} else {
-		// Copy the slice before mutating, in case a client is iterating it:
-		buckets = slices.Clone(buckets)
-		buckets[i] = nil // remove ptr that might be left in the underlying array, for gc
-		buckets = slices.Delete(buckets, i, i+1)
-		bucketRegistry[url] = buckets
+		delete(r.byURL, url)
+		if !bucket.inMemory {
+			bucket._closeAllInstances()
+			delete(r.inMemoryRef, name)
+		}
+		return
 	}
+	// Copy the slice before mutating, in case a client is iterating it:
```

Comment:
Whereabouts do we iterate over the slice without holding the lock?

Reply:
This code was carried over from the original implementation, I've simplified it for refcounting.